core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/rules/auth/TokenAuthenticationRule.scala

```diff
@@ -0,0 +1,96 @@
+import monix.eval.Task
+import org.apache.logging.log4j.scala.Logging
+import tech.beshu.ror.RequestId
+import tech.beshu.ror.accesscontrol.blocks.mocks.MocksProvider
+import tech.beshu.ror.accesscontrol.blocks.rules.Rule
+import tech.beshu.ror.accesscontrol.blocks.rules.Rule.AuthenticationRule.EligibleUsersSupport
+import tech.beshu.ror.accesscontrol.blocks.rules.Rule.RuleResult.{Fulfilled, Rejected}
+import tech.beshu.ror.accesscontrol.blocks.rules.Rule.{AuthenticationRule, RuleName, RuleResult}
+import tech.beshu.ror.accesscontrol.blocks.rules.auth.TokenAuthenticationRule.Settings
+import tech.beshu.ror.accesscontrol.blocks.rules.auth.base.BaseAuthenticationRule
+import tech.beshu.ror.accesscontrol.blocks.rules.auth.base.impersonation.Impersonation
+import tech.beshu.ror.accesscontrol.blocks.rules.auth.base.impersonation.SimpleAuthenticationImpersonationSupport.UserExistence
+import tech.beshu.ror.accesscontrol.blocks.{BlockContext, BlockContextUpdater}
+import tech.beshu.ror.accesscontrol.domain.LoggedUser.DirectlyLoggedUser
+import tech.beshu.ror.accesscontrol.domain.User.Id.UserIdCaseMappingEquality
+import tech.beshu.ror.accesscontrol.domain.{Header, User}
+import tech.beshu.ror.accesscontrol.request.RequestContext
+
+final class TokenAuthenticationRule(val settings: Settings,
+                                    override val impersonation: Impersonation,
+                                    override val caseMappingEquality: UserIdCaseMappingEquality)
+  extends BaseAuthenticationRule
+    with Logging {
+
+  override val name: Rule.Name = TokenAuthenticationRule.Name.name
+
+  override val eligibleUsers: AuthenticationRule.EligibleUsersSupport = EligibleUsersSupport.Available(Set(settings.user))
+
+  override def exists(user: User.Id, mocksProvider: MocksProvider)
+                     (implicit requestId: RequestId,
+                      eq: Eq[User.Id]): Task[UserExistence] = Task.now {
+    if (user === settings.user) UserExistence.Exists
+    else UserExistence.NotExist
+  }
+
+  override protected def tryToAuthenticateUser[B <: BlockContext : BlockContextUpdater](blockContext: B): Task[RuleResult[B]] =
+    Task
+      .unit
+      .map { _ =>
+        val requestContext = blockContext.requestContext
+        if (hasTokenInHeader(requestContext)) {
+          Fulfilled(blockContext.withUserMetadata(_.withLoggedUser(DirectlyLoggedUser(settings.user))))
+        } else {
+          Rejected()
+        }
+      }
+
+  private def hasTokenInHeader(requestContext: RequestContext) = {
+    requestContext
+      .headers
+      .find(_.name === settings.tokenHeader.name)
+      .exists(_.value == settings.tokenHeader.value)
+  }
+}
+
+object TokenAuthenticationRule {
+  implicit case object Name extends RuleName[TokenAuthenticationRule] {
+    override val name: Rule.Name = Rule.Name("token_authentication")
+  }
+
+  final case class Settings(user: User.Id,
+                            tokenHeader: Header)
```

Comment:
I'm not sure if we should store it like that. IMO it's better to have `tokenHeaderName` and `token` fields in the settings. IMO it's easier to grasp